pkg/controller/v1alpha1/routes/route_controller.go

```diff
@@ -0,0 +1,436 @@
+			Labels:    oamServiceLabel,
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeLoadBalancer,
+		},
+	}
+	// assign selector
+	if routeTrait.Spec.Backend != nil && routeTrait.Spec.Backend.SelectLabels != nil {
+		oamService.Spec.Selector = routeTrait.Spec.Backend.SelectLabels
+	}
+
+	port, labels, err := DiscoverPortLabel(workload)
+	if err == nil {
+		if oamService.Spec.Selector == nil {
+			oamService.Spec.Selector = labels
+		}
+		if routeTrait.Spec.Backend == nil {
+			routeTrait.Spec.Backend = &standardv1alpha1.Backend{Port: port}
+		} else if routeTrait.Spec.Backend.Port.String() == "0" {
+			routeTrait.Spec.Backend.Port = port
+		}
+	}
+
+	var servicePort int32 = 443
+	oamService.Spec.Ports = []corev1.ServicePort{
+		{
+			Port:       servicePort,
+			TargetPort: routeTrait.Spec.Backend.Port,
+			Protocol:   corev1.ProtocolTCP,
+		},
+	}
+	// server side apply the service, only the fields we set are touched
+	applyOpts := []client.PatchOption{client.ForceOwnership, client.FieldOwner(routeTrait.GetUID())}
+	if err := r.Patch(ctx, oamService, client.Apply, applyOpts...); err != nil {
+		mLog.Error(err, "Failed to apply to service")
+		return nil, servicePort, err
+	}
+	return &runtimev1alpha1.TypedReference{
+		APIVersion: common.ServiceAPIVersion,
+		Kind:       common.ServiceKind,
+		Name:       oamService.Name,
+		UID:        routeTrait.UID,
+	}, servicePort, nil
+}
+
+// Assume the workload will always having spec.template as PodTemplate if discoverable
+
+func DiscoverPortLabel(workload *unstructured.Unstructured) (intstr.IntOrString, map[string]string, error) {
+	obj, found, _ := unstructured.NestedMap(workload.Object, "spec", "template")
+	if !found {
+		return intstr.IntOrString{}, nil, errors.New("not have spec.template in workload")
+	}
+	data, err := json.Marshal(obj)
+	if err != nil {
+		return intstr.IntOrString{}, nil, fmt.Errorf("convert object err %v", err)
+	}
+	var template corev1.PodTemplate
+	err = json.Unmarshal(data, &template)
+	if err != nil {
+		return intstr.IntOrString{}, nil, fmt.Errorf("convert object to PodTemplate err %v", err)
+	}
+	port := getFirstPort(template.Template.Spec.Containers)
+	if port == 0 {
+		return intstr.IntOrString{}, nil, errors.New("no port found in workload")
+	}
+	return intstr.FromInt(int(port)), template.Labels, nil
+}
+
+func getFirstPort(cs []corev1.Container) int32 {
+	//TODO(wonderflow): exclude some sidecars
+	for _, container := range cs {
+		for _, p := range container.Ports {
+			return p.ContainerPort
+		}
+	}
+	return 0
+}
+
+func (r *Reconciler) createSelfsignedIssuer(ctx context.Context, routeTrait *v1alpha1.Route) (string, error) {
+	var selfSigned = "selfsigned"
+	var issuer = certmanager.Issuer{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: certmanager.SchemeGroupVersion.String(),
+			Kind:       certmanager.IssuerKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      selfSigned,
+			Namespace: routeTrait.Namespace,
```

Comment:
there is a self signed cert in the vela-system, can you use that?

Reply:
that's not cluster-scope, only same namespace can use. So I create one for in each user's namespace